staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/stdlib.go

```diff
@@ -0,0 +1,616 @@
+//
+//     url(<string>) <URL>
+//
+//
+// isURL
+//
+// Returns true if a string is a valid URL.
+//
+//     <string>.isURL() <bool>
+//
+//
+// getScheme / getUsername / getPassword / getHostname / getPort / getPath / getQuery / getFragment
+//
+// Return the parsed components of the URL.
+//
+//     <URL>.getScheme() <string>
+//     <URL>.getUsername() <string>
+//     <URL>.getPassword() <string>
+//     <URL>.getHostname() <string>
+//     <URL>.getPort() <string>
+//     <URL>.getPath() <string>
+//     <URL>.getQuery() <map <string>, <list <string>>
+//     <URL>.getFragment() <string>
+//
+// Examples:
+//     url('https://example.com/').getScheme() // returns 'https'
+//     url('https://user:pass@example.com/').getUsername() // returns 'user'
+//     url('https://user:pass@example.com/').getPassword() // returns 'pass'
+//     url('https://example.com:80/path').getHostname() // returns 'example.com'
+//     url('https://127.0.0.1:80/path').getHostname() // returns '127.0.0.1'
+//     url('https://[::1]:80/path').getHostname() // returns '::1'
+//     url('https://example.com:80/path').getPort() // returns '80'
+//     url('https://example.com/path').getPath() // returns '/path'
+//     url('https://example.com/path?k1=a&k2=b&k2=c').getQuery() // returns { 'k1': ['a'], 'k2': ['b', 'c']}
+//     url('https://example.com/path#overview').getFragment() // returns 'overview'
+//
+//
+func Stdlib() cel.EnvOption {
+	return cel.Lib(stdlib{})
+}
+
+type stdlib struct{}
+
+var paramA = decls.NewTypeParamType("A")
+var typeParamAList = []string{"A"}
+
+func (stdlib) CompileOptions() []cel.EnvOption {
+	return []cel.EnvOption{
+		cel.Declarations(
+			decls.NewFunction("isSorted",
+				decls.NewParameterizedInstanceOverload("list_a_is_sorted_bool",
+					[]*exprpb.Type{decls.NewListType(paramA)},
+					decls.Bool, typeParamAList),
+			),
+			// Only overload sum for a fixed set of list element types
+			// to restrict to a subset of the types of what NewParameterizedInstanceOverload
+			// would allow.
+			decls.NewFunction("sum",
+				decls.NewInstanceOverload("list_int_sum_int",
+					[]*exprpb.Type{decls.NewListType(decls.Int)},
+					decls.Int),
+				decls.NewInstanceOverload("list_uint_sum_uint",
+					[]*exprpb.Type{decls.NewListType(decls.Uint)},
+					decls.Uint),
+				decls.NewInstanceOverload("list_double_sum_double",
+					[]*exprpb.Type{decls.NewListType(decls.Double)},
+					decls.Double),
+				decls.NewInstanceOverload("list_duration_sum_duration",
+					[]*exprpb.Type{decls.NewListType(decls.Duration)},
+					decls.Duration),
+			),
+			decls.NewFunction("max",
+				decls.NewParameterizedInstanceOverload("list_a_max_a",
+					[]*exprpb.Type{decls.NewListType(paramA)},
+					paramA, typeParamAList),
+			),
+			decls.NewFunction("min",
+				decls.NewParameterizedInstanceOverload("list_a_min_a",
+					[]*exprpb.Type{decls.NewListType(paramA)},
+					paramA, typeParamAList),
+			),
```

Comment:
does anything ensure at typecheck-time the items in these lists are of a comparable type (not a list of bools, arrays, or objects)?

Reply:
No, I've added tests showing the behavior. Arrays/maps/objects get a runtime type check error message.I had wanted to apply a `OperandTrait` restriction, but that's only available for type of the 1st argument. I can't attach it to the type param.  This might be addressable in CEL.

Bools, non-intuitively, are comparable in CEL and so are supported.

Short term options are:
- Live with the gap in type checking since the runtime type checker will catch it
- Ditch type params here and list out an overload for every type we want to support. We could swap type params back in if we get the CEL type checker fixed to handle this.

Reply:
If you only want comparable types, then you'll need to list the parameterized list types explicitly for the foreseeable future

Reply:
> Short term options are:
>
> * Live with the gap in type checking since the runtime type checker will catch it
> * Ditch type params here and list out an overload for every type we want to support. We could swap type params back in if we get the CEL type checker fixed to handle this.

I have a slight preference for fully specifying/bounding the types to improve the type checking

Reply:
I'll list them out. It's not that bad.